src/mode/modeHandler.ts

```diff
@@ -603,7 +603,15 @@ export class ModeHandler implements vscode.Disposable {
         this.vimState.alteredHistory = false;
         vimState.historyTracker.ignoreChange();
       } else {
-        vimState.historyTracker.addChange(this.vimState.cursorsInitialState.map(c => c.stop));
+        const addedChange = vimState.historyTracker.addChange(
+          this.vimState.cursorsInitialState.map(c => c.stop)
+        );
+        const mightChangeDocument = recordedState.hasRunOperator
+          ? recordedState.operator.mightChangeDocument
+          : action.mightChangeDocument;
+        if (addedChange && !mightChangeDocument) {
+          throw new Error(`Action '${action.toString()}' changed the document unexpectedly!`);
```

Comment:
This will always throw with #4386 when switching out of Insert mode using `<C-c>` and `<C-o>` because `_isDocumentTextNeeded()` is expecting modes to be different so that the state of the doc after the last change is recorded. When this `addChange()` gets called we will be in Normal Mode and thus `<C-c>` (key=<copy>) might not change the document, which is correct. Typically we would have recorded the changes just before switching out of Insert mode because the `addChange()` in `ModeHandler.handleKeyEventHelper()` takes care of that before the action is ran.

Reply:
Maybe have a `action.willSwitchModes` and pass that over to `addChange()` in `HistoryTracker.handleKeyEventHelper`?

Reply:
I'm tempted to just set `mightChangeDocument` to `true` on these two commands and add to the `mightChangeDocument` comment explaining that since we only want to diff at the end of insert mode, a command that goes from insert mode to normal mode does (sort of) change the document. Thoughts on that?

Reply:
I think `<C-o>` (CommandOneNormalCommandInInsertMode) should definitely have `mightChangeDocument = true` since that's exiting Insert mode. `<C-c>` (CommandEscInsertMode) already has it properly set, but I believe is failing because it gets overwritten to `<copy>`.

Reply:
I guess setting `mightChangeDocument = true` on `<copy>` would solve it.

Reply:
Yeah I like this idea better, would you like me to just commit these changes into #4386?

Reply:
That'd be great, thanks